GAN/main.swift

```diff
@@ -0,0 +1,214 @@
+// Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import TensorFlow
+import Python
+
+// Import Python modules
+let matplotlib = Python.import("matplotlib")
+let np = Python.import("numpy")
+let plt = Python.import("matplotlib.pyplot")
+
+// Turn off using display on server / linux
+matplotlib.use("Agg")
+
+// Some globals
+let epochCount = 10
+let batchSize = 32
+let outputFolder = "./output/"
+let imageHeight = 28, imageWidth = 28
+let imageDim = imageHeight*imageWidth
+let latentDim = 64
+
+func plot(image: Tensor<Float>, name: String) {
+    // Create figure
+    let ax = plt.gca()
+    let array = np.array([image.scalars])
+    let pixels = array.reshape(image.shape)
+    if !FileManager.default.fileExists(atPath: outputFolder) {
+        try! FileManager.default.createDirectory(atPath: outputFolder,
+                            withIntermediateDirectories: false,
+                                             attributes: nil)
+    }
+    ax.imshow(pixels, cmap: "gray")
+    plt.savefig("\(outputFolder)\(name).png", dpi: 300)
+    plt.close()
+}
+
+/// Reads a file into an array of bytes.
+func readFile(_ filename: String) -> [UInt8] {
+    let possibleFolders = [".", "Resources", "GAN/Resources"]
+    for folder in possibleFolders {
+        let parent = URL(fileURLWithPath: folder)
+        let filePath = parent.appendingPathComponent(filename).path
+        guard FileManager.default.fileExists(atPath: filePath) else {
+            continue
+        }
+        let d = Python.open(filePath, "rb").read()
+        return Array(numpy: np.frombuffer(d, dtype: np.uint8))!
+    }
+    print("Failed to find file with name \(filename) in the following folders: \(possibleFolders).")
+    exit(-1)
+}
+
+/// Reads MNIST images and labels from specified file paths.
+func readMNIST(imagesFile: String, labelsFile: String) -> (images: Tensor<Float>,
+                                                           labels: Tensor<Int32>) {
+    print("Reading data.")
+    let images = readFile(imagesFile).dropFirst(16).map { Float($0) }
+    let labels = readFile(labelsFile).dropFirst(8).map { Int32($0) }
+    let rowCount = labels.count
+
+    print("Constructing data tensors.")
+    return (
+        images: Tensor(shape: [rowCount, imageHeight * imageWidth], scalars: images) / 255.0 * 2 - 1,
+        labels: Tensor(labels)
+    )
+}
+
+// Models
```

Comment:
Add an empty line, since this is describing both `Generator` and `Discriminator` below.
```suggestion
// Models

```